fix(linter/react): Fixed false positive with missing key inside React.Children.toArray()

ApollonDeParnasse opened the pull request:
Closes: #4421 

Added three new functions in order to check if an element is inside of React.Children.toArray.
Tests added for this fix are exactly the same as those used by eslint-plugin-react.

ApollonDeParnasse committed:
test(linter/react): Added four new tests

ApollonDeParnasse committed:
fix(linter/react): Allow .map to be used inside React.Children.toArray() without a key

Added three new functions: is_within_children_to_array, is_children
and is_to_array. is_within_children_to_array uses is_children and
is_to_array to check if an element is inside React.Children.toArray().
check_jsx_element calls is_within_children_to_array.

ApollonDeParnasse committed:
chore(linter/react): ran just

mysteryven commented on crates/oxc_linter/src/rules/react/jsx_key.rs:
```diff
@@ -400,6 +444,29 @@ fn test() {
+        r"React.Children.toArray([1, 2 ,3].map(x => <App />));",
     ];
 
     let fail = vec![
```
Could you add a test case `foo.Children.toArray([1, 2 ,3].map(x => <App />));'`? This should fail.

DonIsaac commented:
@mysteryven feel free to merge when you're satisfied.

vHugoObject commented:
The code needs to be updated for the new test but it will finish it ASAP.

DonIsaac commented:
@vHugoObject need some help getting this PR over the finish line?

vHugoObject commented:
Sorry, I got it, I was just busy the last two days.

On Mon, Aug 19, 2024 at 7:16 PM Don Isaac ***@***.***> wrote:

> @vHugoObject <https://github.com/vHugoObject> need some help getting this
> PR over the finish line?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oxc-project/oxc/pull/4945#issuecomment-2297734285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEGOVSWJNABLI2L5RYUSPLZSKDGHAVCNFSM6AAAAABMU3AW7CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOJXG4ZTIMRYGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ApollonDeParnasse committed:
fix(linter/react): Added is_import and import_matcher

is_children now uses is_import to validate imports from React. Now,
this rule will only run if Children and React actually come from the
React module. is_import and import_matcher are actually general
functions that are not at all tied to React.

ApollonDeParnasse committed:
chore(linter/react): ran just

ApollonDeParnasse committed:
Merge branch 'oxc-project:main' into main

vHugoObject commented:
Added the new test and then added new functions so that is_children always verifies imports.

mysteryven commented:
Thank you!
